javascript/ql/src/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -724,14 +697,172 @@ private predicate isSink(DataFlow::Node nd, DataFlow::Configuration cfg, FlowLab
+  |
+    scope = DataFlow::valueNode(container)
+    or
+    container instanceof TopLevel and
+    scope = DataFlow::globalAccessPathRootPseudoNode()
+  )
+}
+
+/**
+ * Holds if `nd` may be reachable from a source under `cfg`, and `context` is
+ * - the most recently used parameter/receiver, or
+ * - the scope of a captured variable that has been used, or
+ * - `globalAccessPathRootPseudoNode`, representing an unrestricted context.
+ *
+ * No store/load matching is done and flow labels are ignored (effectively merged into a single label).
+ */
+private predicate isRelevantForward(
+  DataFlow::Node nd, DataFlow::Configuration cfg, DataFlow::Node context
+) {
+  isSource(nd, cfg, _) and context = DataFlow::globalAccessPathRootPseudoNode()
+  or
+  exists(DataFlow::Node mid |
+    isRelevantForward(mid, cfg, context) and
+    basicExploratoryFlowStep(mid, nd, cfg)
+  )
+  or
+  // Step into a function parameter, while setting the `context` to that parameter.
+  exists(DataFlow::Node arg |
+    isRelevantForward(arg, cfg, _) and
+    (argumentPassing(_, arg, _, nd) or exploratoryBoundInvokeStep(arg, nd)) and
+    context = nd
+  )
+  or
+  // Step through a return if the originating parameter is within the shared environment
+  // of caller and callee (or if no calls have been seen).
+  exists(DataFlow::Node ret, Function f |
+    isRelevantForward(ret, cfg, context) and
+    returnStep(ret, nd, _, f) and
+    getContainerDepthFromInputNode(context) < f.getContainerDepth()
+  )
+  or
+  // Step through a summarized function.
+  exists(DataFlow::Node mid |
+    isRelevantForward(mid, cfg, context) and
+    exploratoryFlowThroughFunction(mid, nd, cfg)
+  )
+  or
+  // When reaching a captured variable, switch `context` to the scope declaring that variable.
+  // This means we we can freely step out to any caller that has the captured variable in scope.
+  isRelevantForward(nd, cfg, _) and
+  capturedVariableScope(nd, context)
+}
+
+/**
+ * Holds if there is a step `pred -> succ` through a function summarized by exploratory flow.
+ */
+pragma[nomagic]
+private predicate exploratoryFlowThroughFunction(
+  DataFlow::Node pred, DataFlow::Node succ, DataFlow::Configuration cfg
+) {
+  exists(DataFlow::InvokeNode invk, Function f, DataFlow::Node returnedParam, DataFlow::Node ret |
+    isRelevantForward(ret, cfg, returnedParam) and
+    returnStep(ret, succ, invk, f) and
+    argumentPassing(invk, pred, f, returnedParam)
+  )
+}
+
+/**
+ * Holds if there exists a forward-relevant sink in any configuration.
+ */
+pragma[nomagic]
+private predicate existsRelevantSink() {
+  exists(DataFlow::Node sink, DataFlow::Configuration cfg |
+    isSink(sink, cfg, _) and
+    isRelevantForward(sink, cfg, _)
+  )
+}
+
+/**
+ * Holds if `pred -> succ` is a step usable by exploratory flow, disregarding call/return
+ * matching.
+ *
+ * Since call/return matching has already been enforced in the forward pass, there is little
```

Comment:
Clarify: can we elaborate on what this little benefit is? Theoretical performance, or is there some subtle semantic property that I fail to see?